### ISSUE TITLE
`callable` type guard removing `Callable` type constraints

### ISSUE THREAD

**Issue by ktbarrett** [astral-sh/ty#3056]
### Summary

I get a false positive for `call-top-callable` in the below code. mypy does not find this code problematic. `Callable[[], int] & Top[(...) -> object]` is `Callable[[], int]` not `Top[(...) -> object]`.

```python
from typing import *

def wew(a: int | Callable[[], int]) -> int:
    if callable(a):
        return a()  # Object of type `Top[(...) -> object]` is not safe to call; its signature is not known
    else:
        return a

```

### Version

0.0.23

**Comment by MeGaGiGaGon** [comment]
This is intended behavior. The issue is that there could be a subclass of `int` that is also `Callable`, which would cause your current code to do the wrong thing:
https://play.ty.dev/6f45552f-e089-4054-afa3-424268eb9696
```py
from typing import *

def wew(a: int | Callable[[], int]) -> int:
    if callable(a):
        return a()  # Object of type `Top[(...) -> object]` is not safe to call; its signature is not known
    else:
        return a

class Sneaky(int):
    def __call__(self) -> str:
        return "Oops"

# Revealed type: `int`
# Runtime type is 'str'
# prints Oops
print(reveal_type(wew(Sneaky(1))))
```

To make your existing code safe, you can invert the guard, which stops the code from accidentally calling an object like `Sneaky`:
https://play.ty.dev/5e35488b-3619-423e-9ddf-ab6882adc2f6
```py
from typing import *

def wew(a: int | Callable[[], int]) -> int:
    if not isinstance(a, int):
        return a()  # No error :)
    else:
        return a
```

Also see https://github.com/astral-sh/ty/issues/1578 for the discussion on if ty should keep the current behavior, as well as https://github.com/astral-sh/ty/issues/2221 for giving a better diagnostic for when a case like this happens.

**Comment by ktbarrett** [comment]
Ah interesting. Unfortunately refactoring the code can involve behavioral changes in the exact case you called out where the two sides are not disjoint.

I think we would like to see a either a setting or some way to describe that the two sides of the union are disjoint, so we are only changing the typing, and not the runtime behavior.

Maybe new syntax to describe this would be helpful. Perhaps this is more of a recommendation for the typing committee.

```python
# comma expresses disjoint union, similar to TypeVar with multiple bound argument vs union bound argument
def wew(a: int, Callable[[], int]) -> int:
    ...

# express intersection with the inverse of the type we wish to be disjoint from
def wew(a: int & ~Callable[..., object] | Callable[[], int]) -> int:
    ...
```

**Comment by AlexWaygood** [comment]
You can already do this if you're willing to use the nonstandardised typeforms in the `ty_extensions` module and you're okay with a lot of "bracketese":

```py
from typing import *
from ty_extensions import Not, Intersection, Top

def wew(a: Intersection[int, Not[Top[Callable[..., object]]]] | Callable[[], int]) -> int:
    if callable(a):
        reveal_type(a)
        return a()  # Object of type `Top[(...) -> object]` is not safe to call; its signature is not known
    else:
        return a

class Sneaky(int):
    def __call__(self) -> str:
        return "Oops"

# error: Argument to function `wew` is incorrect: Expected `(int & ~Top[(...) -> object]) | (() -> int)`, found `Sneaky`
print(reveal_type(wew(Sneaky(1))))
```

however, using `int ~Top[Callable[..., object]]` has the downside that passing a regular `int` to `wew` will now be rejected as well as instances of `Sneaky`:

```py
def test(x: int):
    wew(x)  # Error: Argument to function `wew` is incorrect: Expected `(int & ~Top[(...) -> object]) | (() -> int)`, found `int`
```

you have to explicitly narrow the `int` to the subtype of `int & ~Top[Callable[..., object]]` before ty will allow you to pass it to `wew`:

```py
def test(x: int):
    if not callable(x):
        wew(x)  # fine
```

though even with these revised annotations, ty will still allow you to pass int literals to `wew`, since it knows for a fact that those are never callable:

```py
wew(42)  # fine
```

**Comment by carljm** [comment]
@ktbarrett @AlexWaygood Probably best to follow up in #1578, to centralize the discussion.
